File: old/Privault.Code/Services/WebService.cs
using Privault.Core.Models.DTOs;
using Privault.Core.Services.Contracts;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;

namespace Privault.Core.Services;

/// <inheritdoc cref="IWebService" />
public class WebService : IWebService
{
    // ------------------------------------------------------------------------
    // --- Felder ---
    // ------------------------------------------------------------------------
    
    private readonly ICryptoService _cryptoService;
    private readonly ISessionService _sessionService;
    private readonly IHttpClientFactory _httpClientFactory;

    // ------------------------------------------------------------------------
    // --- Konstruktor ---
    // ------------------------------------------------------------------------
    
    /// <summary>
    /// Initialisiert eine neue Instanz des <see cref="WebService"/>.
    /// </summary>
    /// <param name="cryptoService">Dienst für kryptografische Signaturen und Hashes.</param>
    /// <param name="sessionService">Dienst für den Zugriff auf aktuelle Sitzungsdaten und Schlüssel.</param>
    /// <param name="httpClientFactory">Factory zur Erzeugung von HttpClient-Instanzen.</param>
    /// <exception cref="ArgumentNullException">Wird geworfen, wenn ein Dienst null ist.</exception>
    public WebService(
        ICryptoService cryptoService, 
        ISessionService sessionService,
        IHttpClientFactory httpClientFactory)
    {
        _cryptoService = cryptoService ?? throw new ArgumentNullException(nameof(cryptoService));
        _sessionService = sessionService ?? throw new ArgumentNullException(nameof(sessionService));
        _httpClientFactory = httpClientFactory ?? throw new ArgumentNullException(nameof(httpClientFactory));
    }

    // ------------------------------------------------------------------------
    // --- Öffentliche Methoden ---
    // ------------------------------------------------------------------------

    // --- Resource Version ---
    
    /// <inheritdoc />
    public async Task<VersionResponse> GetServerVersionAsync(string? host = null, string? apiToken = null)
    {
        // WebService aufrufen
        using var client = CreateClient(host, apiToken);
        var response = await client.GetAsync("version");
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Serveranfrage fehlgeschlagen ({response.StatusCode}): {error}");
        }
        var result = await response.Content.ReadFromJsonAsync<VersionResponse>() ?? throw new Exception("Serveranfrage fehlgeschlagen");
        
        // Ergebnis zurückgeben
        return result;
    }
    
    // --- Resource User ---
    
    /// <inheritdoc />
    public async Task<UserResponse> GetUserAsync(string userUuid)
    {
        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.GetAsync($"users/{userUuid}");

        // Antwort auswerten
        if (!response.IsSuccessStatusCode) 
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Serveranfrage fehlgeschlagen ({response.StatusCode}): {error}");
        }
        var result = await response.Content.ReadFromJsonAsync<UserResponse>() ?? throw new Exception("Serveranfrage fehlgeschlagen");
        
        // Ergebnis zurückgeben
        return result;
    }
    
    /// <inheritdoc />
    public async Task<UserResponse?> FindUserAsync(string vaultName, string userName)
    {
        // Daten für Anfrage zusammenstellen
        var vaultHash = _cryptoService.ComputeHash(vaultName);
        var userHash = _cryptoService.ComputeHash(userName);
        
        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.GetAsync($"users?vault_hash={vaultHash}&user_hash={userHash}");

        // Antwort auswerten
        if (!response.IsSuccessStatusCode) 
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Serveranfrage fehlgeschlagen ({response.StatusCode}): {error}");
        }
        var result = await response.Content.ReadFromJsonAsync<UserResponse>();
        
        // Ergebnis zurückgeben
        return result;
    }
    
    /// <inheritdoc />
    public async Task<UserResponse> RegisterUserAsync(string vaultName, string userName)
    {
        // Daten für Anfrage zusammenstellen
        var user = _sessionService.User ?? throw new Exception("Kein lokaler Benutzer geladen.");
        var settings = _sessionService.Settings ?? throw new Exception("Keine Tresor-Einstellungen geladen.");
        var body = new
        {
            user_uuid = user.Uuid,
            vault_hash = _cryptoService.ComputeHash(vaultName),
            user_hash = _cryptoService.ComputeHash(userName),
            salt = settings.Salt,
            public_key = user.PublicKey,
            encrypted_private_key = settings.EncryptedPrivateKey
        };

        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.PostAsJsonAsync("users", body);
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Registrierung fehlgeschlagen: {response.StatusCode} - {error}");
        }
        var result = await response.Content.ReadFromJsonAsync<UserResponse>() ?? throw new Exception("Serveranfrage fehlgeschlagen");
        
        // Ergebnis zurückgeben
        return result;
    }

    /// <inheritdoc />
    public async Task ChangePasswordAsync(string userUuid, string salt, string encryptedPrivateKey)
    {
        // Daten für Anfrage zusammenstellen
        var body = new { salt, encrypted_private_key = encryptedPrivateKey };
        
        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.PutAsJsonAsync($"users/{userUuid}/password", body);
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Server-Update fehlgeschlagen: {response.StatusCode} - {error}");
        }
    }

    /// <inheritdoc />
    public async Task SaveFriendsAsync(string userUuid, string encryptedFriends)
    {
        // Daten für Anfrage zusammenstellen
        var body = new { encrypted_friends = encryptedFriends };
        
        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.PutAsJsonAsync($"users/{userUuid}/friends", body);
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Friends Push fehlgeschlagen ({response.StatusCode}): {error}");
        }
    }

    /// <inheritdoc />
    public async Task<List<PublicKeyResponse>> GetPublicKeysAsync(string userUuid)
    {
        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.GetAsync($"users/{userUuid}/public_keys");
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Friends Pull fehlgeschlagen ({response.StatusCode}): {error}");
        }
        var result = await response.Content.ReadFromJsonAsync<List<PublicKeyResponse>>() ?? [];
        
        // Ergebnis zurückgeben
        return result;
    }
    
    // --- Bulk-Aktion Sync ---
    
    /// <inheritdoc />
    public async Task<SyncPullResponse> PullSyncAsync(string userUuid, DateTime since)
    {
        // Daten für Anfrage zusammenstellen
        var utcSince = since.Kind == DateTimeKind.Utc ? since : since.ToUniversalTime();
        
        // WebService aufrufen
        using var client = CreateClient();
        // ":O" (Round-trip) erzeugt den ISO-8601-String (mit 'Z' am Ende)
        var response = await client.GetAsync($"users/{userUuid}/entries/sync?since={utcSince:O}");
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Synchronisation fehlgeschlagen ({response.StatusCode}): {error}");
        }
        var result = await response.Content.ReadFromJsonAsync<SyncPullResponse>() ?? throw new Exception("Serveranfrage fehlgeschlagen");
        
        // Ergebnis zurückgeben
        return result;
    }

    /// <inheritdoc />
    public async Task PushSyncAsync(string userUuid, SyncPushRequest request)
    {
        // Daten für Anfrage zusammenstellen
        var json = JsonSerializer.Serialize(request);
        var body = new StringContent(json, Encoding.UTF8, "application/json");
        
        // WebService aufrufen
        using var client = CreateClient();
        // System.Text.Json Serialize, damit DateTime im DTO als ISO-8601 übertragen wird
        var response = await client.PostAsync($"users/{userUuid}/entries/sync", body);
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Push fehlgeschlagen ({response.StatusCode}): {error}");
        }
    }

    // --- Resource Attachment ---

    /// <inheritdoc />
    public async Task<AttachmentResponse> DownloadAttachmentAsync(string attachmentUuid)
    {
        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.GetAsync($"attachments/{attachmentUuid}");
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Download fehlgeschlagen ({response.StatusCode}): {error}");
        }
        var result = await response.Content.ReadFromJsonAsync<AttachmentResponse>() ?? throw new Exception("Serveranfrage fehlgeschlagen");
        
        // Ergebnis zurückgeben
        return result;
    }

    /// <inheritdoc />
    public async Task UploadAttachmentAsync(string entryUuid, string attachmentUuid, string encryptedMetaBase64, string encryptedContentBase64)
    {
        // Daten für Anfrage zusammenstellen
        var body = new 
        { 
            entry_uuid = entryUuid, 
            encrypted_meta = encryptedMetaBase64, 
            encrypted_content = encryptedContentBase64 
        };

        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.PutAsJsonAsync($"attachments/{attachmentUuid}", body);
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Attachment-Upload fehlgeschlagen ({response.StatusCode}): {error}");
        }
    }

    // -- Resource Vault --
    
    /// <inheritdoc />
    public async Task CleanTestAsync(string vaultName)
    {
        // Daten für Anfrage zusammenstellen
        var vaultHash = _cryptoService.ComputeHash(vaultName);
        
        // WebService aufrufen
        using var client = CreateClient();
        var response = await client.DeleteAsync($"vaults?vault_hash={vaultHash}");
        
        // Antwort auswerten
        if (!response.IsSuccessStatusCode)
        {
            var error = await response.Content.ReadAsStringAsync();
            throw new Exception($"Test-Cleanup fehlgeschlagen ({response.StatusCode}): {error}");
        }
    }

    // ------------------------------------------------------------------------
    // --- Private Methoden ---
    // ------------------------------------------------------------------------
    
    /// <summary>
    /// Erzeugt einen vorkonfigurierten HttpClient mit API-Token und RSA-Sicherheitsheadern.
    /// <param name="host">Die Basis-URL des Servers.</param>
    /// <param name="apiToken">Das API-Token zur Authentifizierung.</param>
    /// </summary>
    private HttpClient CreateClient(string? host = null, string? apiToken = null)
    {
        // Aus den aktuellen Settings der Session (DB)
        var user = _sessionService.User;
        var privKey = _sessionService.PrivateKey;
        host ??= _sessionService.Settings?.Host ?? string.Empty; 
        apiToken ??= _sessionService.Settings?.ApiToken ?? string.Empty;

        // Validierung der URL und API-Token
        if (string.IsNullOrWhiteSpace(host))
            throw new Exception("Kein Sync-Server konfiguriert. Bitte überprüfe die Einstellungen.");
        if (!Uri.TryCreate(host, UriKind.Absolute, out _))
            throw new Exception($"Sync-Server '{host}' ist ungültig. Bitte überprüfe die Einstellungen.");
        if (string.IsNullOrWhiteSpace(apiToken))
            throw new Exception("Kein API-Token konfiguriert. Bitte überprüfe die Einstellungen.");

        // Client über Factory erzeugen
        var client = _httpClientFactory.CreateClient("PrivaultApi");
        client.BaseAddress = new Uri(host.EndsWith('/') ? host : host + "/");

        // Authentifizierung (Bearer)
        client.DefaultRequestHeaders.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", apiToken);
        client.DefaultRequestHeaders.Add("X-API-Token", apiToken); // todo Bearer sollte allein reichen

        // RSA Authentifizierung
        if (user != null && privKey != null && !string.IsNullOrEmpty(user.Uuid))
        {
            // Wir signieren die Kombination aus UUID und einem aktuellen Zeitstempel (gegen Replay-Attacks)
            var timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString();
            var payload = $"{user.Uuid}:{timestamp}";
                
            // Signatur erstellen (CryptoService nutzen)
            var signature = _cryptoService.SignData(Encoding.UTF8.GetBytes(payload), privKey);
                
            // Sicherstellen, dass die Signatur für Header valide ist (keine Zeilenumbrüche)
            signature = signature.Replace("\r", "").Replace("\n", "").Trim();

            client.DefaultRequestHeaders.Add("X-User-Uuid", user.Uuid);
            client.DefaultRequestHeaders.Add("X-Timestamp", timestamp);
            client.DefaultRequestHeaders.Add("X-Signature", signature);
        }

        // Xdebug-Cookie (für Server-Debugging; non-blocking)
//#if DEBUG
        client.DefaultRequestHeaders.Add("Cookie", "XDEBUG_SESSION=PHPSTORM");
//#endif

        return client;
    }
}

File: lib/services/web_service.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'package:privault/core/app_error.dart';
import 'package:privault/models/dtos/sync_dtos.dart';
import 'package:privault/models/dtos/user_response.dart';
import 'package:privault/models/dtos/version_response.dart';
import 'package:privault/services/crypto_service.dart';

/// Dienst für die Kommunikation mit der priVault-API (Backend).
/// Kümmert sich um Authentifizierung (API-Token) und Autorisierung (RSA-Signatur der Requests).
class WebService {

  // ------------------------------------------------------------------------
  // --- Verwendete Dienste (Abhängigkeiten) ---
  // ------------------------------------------------------------------------

  final CryptoService _cryptoService;

  // ------------------------------------------------------------------------
  // --- Interne Variablen ---
  // ------------------------------------------------------------------------

  final Dio _dio;
  String _apiToken = '';
  String? _userUuid;
  Uint8List? _privateKeyBytes;

  // ------------------------------------------------------------------------
  // --- Initialisierung / Lifecycle ---
  // ------------------------------------------------------------------------

  /// Konstruktor.
  ///
  /// [host] ist die Serveradresse ohne "/api".
  /// [dio] kann für Unit-Tests übergeben werden, um Header oder Mocks zu injizieren.
  WebService(this._cryptoService, {String? host, String? apiToken, Dio? dio})
      : _dio = dio ?? Dio() {

    _apiToken = apiToken ?? '';

    // Falls dio von außen kommt, stellen wir sicher, dass die BaseUrl passt
    if (host != null && host.isNotEmpty) {
      _dio.options.baseUrl = host.endsWith('/') ? '${host}api/' : '$host/api/';
    }

    // Default Headers (API Token & Debug Cookie)
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) {
          options.headers['Authorization'] = 'Bearer $_apiToken';
          options.headers['X-API-Token'] = _apiToken;
          options.headers['Content-Type'] = 'application/json';
          options.headers['Accept'] = 'application/json';
          if (kDebugMode) {
            options.headers['Cookie'] = 'XDEBUG_SESSION=PHPSTORM';
          }
          return handler.next(options);
        },
      ),
    );

    // Signature Interceptor
    // Erzeugt dynamisch für jeden Request eine frische RSA-Signatur
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) async {
          if (_userUuid != null && _privateKeyBytes != null) {
            // Wir signieren die Kombination aus UUID und einem aktuellen Zeitstempel (gegen Replay-Attacks)
            final timestamp = (DateTime.now().toUtc().millisecondsSinceEpoch ~/ 1000).toString();
            final payload = '$_userUuid:$timestamp';

            try {
              // Signatur erstellen (CryptoService nutzen)
              var signature = await _cryptoService.signData(Uint8List.fromList(utf8.encode(payload)), _privateKeyBytes!);

              // Sicherstellen, dass die Signatur für Header valide ist (keine Zeilenumbrüche)
              options.headers['X-User-Uuid'] = _userUuid;
              options.headers['X-Timestamp'] = timestamp;
              options.headers['X-Signature'] = signature.replaceAll('\r', '').replaceAll('\n', '').trim();
            } catch (e) {
              debugPrint("❌ [SIGN] FEHLER IM INTERCEPTOR: $e");
            }
          }
          return handler.next(options);
        },
      ),
    );

    // Log-Ausgabe für Debug-Zwecke
    if (kDebugMode) {
      _dio.interceptors.add(
        LogInterceptor(
          request: true,
          requestHeader: true,
          requestBody: true,
          responseHeader: true,
          responseBody: true,
          error: true,
          logPrint: (obj) => debugPrint(obj.toString()),
        ),
      );
    }
  }

  // --- Konfiguration ---

  /// Aktualisiert die Konfiguration des WebServices.
  /// [host] ist die Serveradresse ohne "/api".
  void updateConfig({required String host, required String apiToken}) {
    if (host.isNotEmpty) {
      _dio.options.baseUrl = host.endsWith('/') ? '${host}api/' : '$host/api/';
    }
    _apiToken = apiToken;
  }

  /// Aktualisiert die Signature-Daten des WebServices.
  void setSignatureData({required String userUuid, required Uint8List privateKey, String? publicKey}) {
    _userUuid = userUuid;
    _privateKeyBytes = privateKey;
  }

  /// Löscht die Signature-Daten des WebServices.
  void clearSignatureData() {
    _userUuid = null;
    _privateKeyBytes = null;
  }

  // ------------------------------------------------------------------------
  // --- Methoden bzgl. Resource Version ---
  // ------------------------------------------------------------------------

  /// Fragt die aktuelle API-Version des Servers ab.
  ///
  /// Falls die Serverantwort ein unerwartetes Format hat, wird
  /// `VersionResponse` mit leeren Werten zurückgegeben.
  Future<VersionResponse> getServerVersion() async {
    final response = await _dio.get('version');
    return VersionResponse.fromJson(response.data);
  }

  // ------------------------------------------------------------------------
  // --- Methoden bzgl. Resource User ---
  // ------------------------------------------------------------------------

  /// Sucht im angegebenen Tresor nach einem bestimmten Benutzer.
  Future<UserResponse?> findUser(String vaultName, String userName) async {
    final vaultHash = _cryptoService.computeHash(vaultName);
    final userHash = _cryptoService.computeHash(userName);
    final response = await _dio.get(
      'users',
      queryParameters: {
        'vault_hash': vaultHash,
        'user_hash': userHash,
      },
    );
    if (response.data == null || response.data.toString().isEmpty) return null;
    return UserResponse.fromJson(response.data);
  }

  /// Registriert einen neuen Benutzer auf dem Server.
  Future<UserResponse> registerUser({required String vaultName, required String userName, required String userUuid, required String salt, required String publicKey, required String encryptedPrivateKey, required DateTime masterKeyTimestamp}) async {
    final response = await _dio.post(
      'users',
      data: {
        'user_uuid': userUuid,
        'vault_hash': _cryptoService.computeHash(vaultName),
        'user_hash': _cryptoService.computeHash(userName),
        'salt': salt,
        'public_key': publicKey,
        'encrypted_private_key': encryptedPrivateKey,
        'master_key_timestamp': masterKeyTimestamp.toUtc().toIso8601String(),
      },
    );
    return UserResponse.fromJson(response.data);
  }

  /// Überträgt eine Passwortänderung (neues Salt und verschlüsselter Private Key) zum Server.
  Future<void> changePassword(String userUuid, String salt, String encryptedPrivateKey, DateTime masterKeyTimestamp) async {
    await _dio.put(
      'users/$userUuid/password',
      data: {
        'salt': salt,
        'encrypted_private_key': encryptedPrivateKey,
        'master_key_timestamp': masterKeyTimestamp.toUtc().toIso8601String(),
      },
    );
  }

  /// Ruft die öffentlichen RSA-Schlüssel aller Benutzer des Tresors ab.
  Future<List<Map<String, dynamic>>> getPublicKeys(String userUuid) async {
    final response = await _dio.get('users/$userUuid/public_keys');
    return List<Map<String, dynamic>>.from(response.data);
  }

  /// Speichert die verschlüsselte Freundesliste des Benutzers auf dem Server.
  Future<void> saveFriends(String userUuid, String encryptedFriends) async {
    await _dio.put('users/$userUuid/friends', data: {'encrypted_friends': encryptedFriends});
  }

  /// Ändert den Benutzernamen auf dem Server.
  ///
  /// Wird beim Sync aufgerufen, wenn [user.name] != [user.syncedName].
  Future<void> patchUserName(String userUuid, String userName) async {
    await _dio.patch(
      'users/$userUuid/name',
      data: {
        'user_hash': _cryptoService.computeHash(userName),
      },
    );
  }

  // ------------------------------------------------------------------------
  // --- Methoden bzgl. Bulk-Aktion Sync ---
  // ------------------------------------------------------------------------

  /// Ruft alle Änderungen seit der letzten Synchronisation ab.
  ///
  /// `since` ist Zeitpunkt des letzten erfolgreichen Abgleichs (UTC).
  Future<SyncPullResponse> pullSync(String userUuid, DateTime since) async {
    // ":O" in C# erzeugt einen ISO-8601 String. Das Äquivalent in Dart ist toIso8601String().
    final response = await _dio.get('users/$userUuid/entries/sync', queryParameters: {'since': since.toUtc().toIso8601String()});
    return SyncPullResponse.fromJson(response.data);
  }

  /// Überträgt lokale Änderungen zum Server.
  ///
  /// `request` ist das Payload mit den zu synchronisierenden Daten.
  Future<void> pushSync(String userUuid, SyncPushRequest request) async {
    await _dio.post('users/$userUuid/entries/sync', data: request.toJson());
  }

  // ------------------------------------------------------------------------
  // --- Methoden bzgl. Resource Attachment ---
  // ------------------------------------------------------------------------

  /// Lädt die verschlüsselten Daten eines Dateianhangs vom Server herunter.
  Future<Map<String, dynamic>> downloadAttachment(String attachmentUuid) async {
    final response = await _dio.get('attachments/$attachmentUuid');
    return response.data as Map<String, dynamic>;
  }

  /// Lädt einen neuen oder geänderten Anhang zum Server hoch.
  ///
  /// `encryptedContent` ist Base64-kodiert.
  Future<void> uploadAttachment(String entryUuid, String attachmentUuid, String encryptedMeta, String encryptedContent) async {
    await _dio.put(
      'attachments/$attachmentUuid',
      data: {
        'entry_uuid': entryUuid,
        'encrypted_meta': encryptedMeta,
        'encrypted_content': encryptedContent,
      },
    );
  }

  // ------------------------------------------------------------------------
  // --- Methoden bzgl. Resource Vault ---
  // ------------------------------------------------------------------------

  /// Räumt den Test-Tresor serverseitig auf (DELETE /test?vault_hash=...).
  ///
  /// Wird typischerweise von Integrationstests verwendet.
  Future<void> cleanTest(String vaultName) async {
    final vaultHash = _cryptoService.computeHash(vaultName);
    await _dio.delete('vaults', queryParameters: {'vault_hash': vaultHash});
  }

  /// Löscht den Tresor des Benutzers serverseitig (RSA-geschützt).
  ///
  /// Entspricht der Option "Nur auf dem Server löschen" im Einstellungen-Dialog.
  Future<void> deleteVault(String userUuid) async {
    await _dio.delete('users/$userUuid/vault');
  }

  // ------------------------------------------------------------------------
  // --- Fehlerauswertung ---
  // ------------------------------------------------------------------------

  /// Wandelt den Verbindungsfehler in ein FormError um.
  static AppError convertDioError(DioException de) {
    String message;
    ErrorCode code = ErrorCode.networkError;
    final host = de.requestOptions.baseUrl;

    switch (de.type) {
      case DioExceptionType.connectionError:
        message = 'Service auf $host nicht erreicht. Prüfe deine Internetverbindung und die Serveradresse.';
        break;

      case DioExceptionType.connectionTimeout:
      case DioExceptionType.sendTimeout:
      case DioExceptionType.receiveTimeout:
        message = 'Server $host antwortet nicht rechtzeitig. Ist er online?';
        break;

      case DioExceptionType.badCertificate:
        message = 'Das Sicherheitszertifikat des Servers $host ist ungültig oder abgelaufen.';
        break;

      case DioExceptionType.badResponse: // z.B. kein JSON
        final status = de.response?.statusCode;
        if (status == 404) {
          code = ErrorCode.noSyncService;
          message = 'Auf dem Server $host läuft kein PriVault Sync-Service. Überprüfe die Serveradresse.';
        } else if (status == 401) {
          code = ErrorCode.unauthorized;
          message = 'Der API-Token wurde vom Server abgelehnt.';
        } else if (status == 503) {
          message = 'Der Server $host kann vorübergehend keine Anfrage bearbeiten.';
        } else { // Fallback
          final msg = de.response?.statusMessage ?? (de.message ?? 'Serverfehler');
          message = status != null ? '$msg (Code $status)' : msg;
        }
        break;

      default:
        // Fallback
        final msg = de.response?.statusMessage ?? (de.message ?? 'Netzwerkfehler');
        message = de.response?.statusCode != null ? '$msg (Code ${de.response?.statusCode})' : msg;
    }

    return AppError(code, text: message);
  }
}